cookbooks/aws-parallelcluster-test/recipes/test_nvidia.rb

```
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+#
+# Cookbook:: aws-parallelcluster-test
+# Recipe:: test_nvidia
+#
+# Copyright:: 2013-2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+return if node['cluster']['nvidia']['enabled'] != 'yes' && node['cluster']['nvidia']['enabled'] != true
+
+bash "check Nvidia drivers" do
+  cwd Chef::Config[:file_cache_path]
+  code <<-TEST
+    driver_ver="#{node['cluster']['nvidia']['driver_version']}"
+    export PATH="/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin"
+
+    echo "Testing Nvidia driver version"
+    driver_output=$(nvidia-smi | grep -E -o "Driver Version: [0-9.]+")
+    [[ "${driver_output}" != "Driver Version: ${driver_ver}" ]] && "ERROR Installed version ${driver_output} but expected ${driver_ver}" && exit 1
+    echo "Correctly installed Nvidia ${driver_output}"
+
+    if [[ "#{node['conditions']['efa_supported']}" == "true" ]]; then
```

Double check if we should remove this check because EFA could be installed at runtime

---

EFa is installed at installation time (install recipe not config recipe), so the check makes sense here.

---

but if it's installed at AMI built time, I expect the check is already done in build/test recipes

---

The tests you're referring to are executed at the end of the installation phase.
Kitchen tests are executed at the end of the configuration phase, so they basically test the final result.
If this is the case I think it makes sense to check it because we are testing the final outcome of the overall process with very limited cost, so with high ROI. What do you think?

---

cost is not very limited, I think it does have a considerable impact on test duration.
The underlying logic is that kitchen tests are meant to test changes applied during configuration, so I don't think we should check again all the features and setup if those aren't changed during configuration phase.
Otherwise when we create a generic test, it means that we need to check everything again from scratch, while instead it should be that a test is meant to check a specific phase so we can fail fast on the process.

---

Agree on your proposal